src/Array.hpp

```diff
@@ -64,40 +59,40 @@ namespace LvArray
  *         one template argument that describes the type of the data being stored (T).
  */
 template< typename T,
-          int NDIM,
-          typename PERMUTATION=camp::make_idx_seq_t< NDIM >,
-          typename INDEX_TYPE=std::ptrdiff_t,
-          template< typename > class BUFFER_TYPE=NewChaiBuffer >
+          typename PERMUTATION,
+          typename INDEX_TYPE,
+          template< typename > class BUFFER_TYPE >
```

Comment:
So I removed the `NDIM` template argument because the permutation carries the dimension as well and by removing the default template arguments you'll always have to specify the permutation anyways. I'm now having second thoughts about doing this for the following reasons

- Both `ArrayView` and `ArraySlice` have `NDIM` as the second template parameter and this breaks the symmetry.
- If you do the following
    ```c++
    template< typename T, typename PERMUTATION=camp::make_idx_seq_t< 2 > >
    using array2d = Array< T, PERMUTATION, std::ptrdiff_t, NewChaiBuffer >
    ```
    the following is valid `array2d< int, RAJA::PERM_I > really_a_1d_array;`
- It makes overloading on dimension harder
    ```c++
    // Operate on a 1D array.
    void operate( Array< int, 1, RAJA::PERM_I, std::ptrdiff_t, NewChaiBuffer > const & x );

    // Operate on a 2D array.
    template< typename PERM >
    void operate( Array< int, 2, PERM, std::ptrdiff_t, NewChaiBuffer > const & x );
    ```
    becomes
    ```c++
    // Operate on a 1D array.
    template< typename PERM >
    std::enable_if_t< getDimension( PERM ) == 1 >
    operate( Array< int, RAJA::PERM_I, std::ptrdiff_t, NewChaiBuffer > const & x );

    // Operate on a 2D array.
    template< typename PERM >
    std::enable_if_t< getDimension( PERM ) == 2 >
    void operate( Array< int, RAJA::PERM_I, std::ptrdiff_t, NewChaiBuffer > const & x );
    ```
@rrsettgast  thoughts?

Reply:
I think keeping `NDIM` in the parameter list makes it much clearer what is happening....for the reasons that you list, and probably some other ones we haven't encountered yet. The redundancy isn't the worst thing in the world. Does `NDIM` help to check that you specified an appropriate permutation?

Reply:
Yes, with `NDIM` in the template list `Array< int, 2, RAJA::PERM_I >` would fail with a static_assert. I'll add it back in.

Reply:
Done.